### DISCUSSION TITLE
How to send/recv data after dcutr ConnectionEstablished

### DISCUSSION THREAD

**Discussion by simon-fu** [libp2p/rust-libp2p#4440] [General]
I run the dcutr example with relay-server , and both of client nodes receive ConnectionEstablished  event successfully. Next step I want to send something like "hello world" from one client to the other. But I don't know how to do that. I can't find any function in dcutr::Behaviour for sending/receiving data.

Could anyone told me how to do that?

Further more, Can anyone provide some guidance or examples on how to use yamux to create multiple streams after establishing a connection with DCUTR?

Thank you in advance!

**Comment by thomaseizinger** [comment]
> Further more, Can anyone provide some guidance or examples on how to use yamux to create multiple streams after establishing a connection with DCUTR?

Yamux is handled implicitly for you, you don't need to deal with it.

> Could anyone told me how to do that?

You need to create a separate protocol for that! See https://github.com/libp2p/rust-libp2p/tree/master/examples/file-sharing for some point-to-point communication using `libp2p-request-response`.

**Reply by simon-fu** [reply]
Thanks for the reply!
Yes , I do read the file-sharing example, but I really don't know how to do next.

Could it work if I add a field with type request_response::cbor::Behaviour to the dcutr's Behaviour like this:
```
    #[derive(NetworkBehaviour)]
    #[behaviour(to_swarm = "Event")]
    struct Behaviour {
        relay_client: relay::client::Behaviour,
        ping: ping::Behaviour,
        identify: identify::Behaviour,
        dcutr: dcutr::Behaviour,
        request_response: request_response::cbor::Behaviour<String, String>,
    }
```

and after ConnectionEstablished send data like this:
```
swarm
.behaviour_mut()
.request_response
.send_request(&peer, "hello world".into());
```

Is it right?

**Reply by thomaseizinger** [reply]
Yep that looks about right :)

**Reply by simon-fu** [reply]
Thank you again. 
There is last question that request_response mod implements request/response protocol , and which mod implements stream protocol   just like TcpStream ？

**Reply by mxinden** [reply]
> There is last question that request_response mod implements request/response protocol , and which mod implements stream protocol just like TcpStream ？

None today. There have been plans https://github.com/libp2p/rust-libp2p/pull/1947 but they never merged into master.

**Reply by simon-fu** [reply]
It is sad to hear that.  Stream transport is widely used and it would be very helpful if rust-libp2p supported that。

**Reply by thomaseizinger** [reply]
We support streams (they are the primitive that everything builds on top of), you just need to implement your protocol in terms of `ConnectionHandler` and `NetworkBehaviour`.

**Reply by simon-fu** [reply]
Got it. I will try to implement it later directly based on ConnectionHandler and NetworkBehaviour. It is a little difficult for  me to do that since I am newbie to rust-libp2p. It would be very helpful if rust-libp2p supported that natively.

**Reply by thomaseizinger** [reply]
> It is a little difficult for  me to do that since I am newbie to rust-libp2p.

There is a bit of a learning curve to it which we are actively working on reducing. But I think that same as Rust itself, it is worth it :)

> It would be very helpful if rust-libp2p supported that natively.

We always welcome contributions of new modules!
